### PR TITLE
Recognized fma methods on Power

### PR DESCRIPTION
**Pull request by wbh123456**
Add the following fma methods into J9 Recognized Methods on Power. Also added fma tests cases to test these recognized methods.
- `TR::java_lang_Math_fma_D`
-  `TR::java_lang_Math_fma_F`
- `TR::java_lang_StrictMath_fma_D` 
- `TR::java_lang_StrictMath_fma_F`

Implement the methods in function `J9::Power::CodeGenerator::inlineDirectCall()`.
Add 2 helper functions:
- `static TR::Register *inlineDoublePrecisionFPTrg1Src3(TR::Node *node, TR::InstOpCode::Mnemonic op, TR::CodeGenerator *cg)`
	--> Return the target register generated by `op` and `node` with 3 children in Double Precision
- `static TR::Register *inlineSinglePrecisionFPTrg1Src3(TR::Node *node, TR::InstOpCode::Mnemonic op, TR::CodeGenerator *cg)`
	--> Return the target register generated by `op` and `node` with 3 children in Single Precision

Issue: #7474
Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

### CODE REVIEW

**Comment by wbh123456** [comment]
Performed and passed extended internal testings.

**Comment by wbh123456** [comment]
FYI @0xdaryl

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -10017,6 +10017,27 @@ static TR::Register *inlineSinglePrecisionFPTrg1Src2(TR::Node *node, TR::InstOpC
    return targetRegister;
    }
 
+static TR::Register *inlineSinglePrecisionFPTrg1Src3(TR::Node *node, TR::InstOpCode::Mnemonic op, TR::CodeGenerator *cg)
+   {
+   TR_ASSERT(node->getNumChildren() == 3, "Wrong number of children in inlineSinglePrecisionFPTrg1Src3");
```
Comment (andrewcraik) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: should be a fatal assert

Comment (0xdaryl) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Yes.  Arguably, the others should be fatal asserts as well.

Comment (wbh123456) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Changed only my implementation from `TR_ASSERT` to `TR_ASSERT_FATAL` in https://github.com/eclipse/openj9/commit/66ab479e09f9e9987a63853c41547da3095bce75. The `assert` in other functions remains as just `TR_ASSERT`.

**Comment by wbh123456** [comment]
> This change adds support for recognizing the methods, but does not include changes to tell the inliner to not inline these methods. You need to add the method to the appropriate list in J9::Power::CodeGenerator::suppressInliningOfRecognizedMethod in
> 
> https://github.com/eclipse/openj9/blob/17375491a028ab757e49285ef3d7d736988c9b15/runtime/compiler/p/codegen/J9CodeGenerator.cpp#L285

Fixed in 58d9ffd

**Comment by wbh123456** [comment]
#### Things Need to be checked:
- [ ] Whether the `fma` instruction opcodes are always available on Power architecture (e.g. is there going to be if/else flag checks on the architecture before using `fma` instruction opcodes). 
*If not, then need to conditionally suppress the inlining of `Math.fma` method only if ` fma` is available on the architecture.*
- [ ] Whether the `fma` instruction opcodes perform the same way as java `math.fma()` implementation. 
*There are boundary conditions and corner cases in fma operation, which is properly defined in defined in IEEE 754-2008*

**Comment by wbh123456** [comment]
> #### Things Need to be checked:
> * [x]  Whether the `fma` instruction opcodes perform the same way as java `math.fma()` implementation.
>   _There are boundary conditions and corner cases in fma operation, which is properly defined in defined in IEEE 754-2008_
> 
> > The definitions of `Power::fmadd`/`Power::fmadds` follow the definition of java `Math.fma` method. Also added fma test cases to ensure the recognized method `fma` behaved the same as `fma` Java implementation. [957dfaf](https://github.com/eclipse/openj9/commit/957dfaf079ffd4d08c18726d225c9eafa7ac1765)
> 
> * [ ]  Whether the `fma` instruction opcodes are always available on Power architecture (e.g. is there going to be if/else flag checks on the architecture before using `fma` instruction opcodes).
>   _If not, then need to conditionally suppress the inlining of `Math.fma` method only if ` fma` is available on the architecture._

**Comment by wbh123456** [comment]
> * [ ]  Whether the `fma` instruction opcodes are always available on Power architecture (e.g. is there going to be if/else flag checks on the architecture before using `fma` instruction opcodes).
>   _If not, then need to conditionally suppress the inlining of `Math.fma` method only if ` fma` is available on the architecture._

@gita-omr @zl-wang . May I ask your opinions on this question? Do you know if `fma` instruction opcode on Power (e.g. `fmadd` / `fmadds`) are always available or is there a proper way to check their availability on Power? @0xdaryl FYI.

**Comment by zl-wang** [comment]
fma is always available on processors we supported.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp [hunk 2]
```diff
@@ -9837,6 +9858,27 @@ static TR::Register *inlineDoublePrecisionFPTrg1Src2(TR::Node *node, TR::InstOpC
    return targetRegister;
    }
 
+static TR::Register *inlineDoublePrecisionFPTrg1Src3(TR::Node *node, TR::InstOpCode::Mnemonic op, TR::CodeGenerator *cg)
```
Comment (0xdaryl) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Can you combine `inlineSinglePrecisionFPTrg1Src3` and this function into a single function to save .so size?  The code is identical except for the size of the target register.

File: test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+        AssertJUnit.assertEquals(r, expected);
+
+        a = min * 2;
+        b = min * 2;
+        c = min * 2;
+        r = Math.fma(-a, -b, -c);
+        expected = -2.8E-45f;
+        AssertJUnit.assertEquals(r, expected);
+        }
+}
```
Comment (0xdaryl) [inline test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java]: Add EOL

File: test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java [hunk 4]
```diff
@@ -0,0 +1,549 @@
+/*******************************************************************************
+ * Copyright (c) 2001, 2019 IBM Corp. and others
```
Comment (0xdaryl) [inline test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java]: These are all new tests, aren't they?  In that case, shouldn't the first copyright date be 2019?

Comment (0xdaryl) [inline test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java]: I'm not sure if there is a file naming scheme in place for these tests, but I would prefer `Test_Math_fma.java` for this file and `Test_StrictMath_fma.java` for the other.  cc @smlambert in case she has some guidance to offer.

File: test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+        AssertJUnit.assertTrue(Double.isNaN(r));
+
+        a = -Math.random();
+        b = Double.POSITIVE_INFINITY; 
+        c = Double.POSITIVE_INFINITY;
+        r = Math.fma(a, b, c);
+        AssertJUnit.assertTrue(Double.isNaN(r));
+
+        /**
+            * Testing infinity numbers:
```
Comment (0xdaryl) [inline test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java]: Please fix formatting.

File: test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma_Strict.java [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+        AssertJUnit.assertEquals(r, expected);
+
+        a = min * 2;
+        b = min * 2;
+        c = min * 2;
+        r = StrictMath.fma(-a, -b, -c);
+        expected = -2.8E-45f;
+        AssertJUnit.assertEquals(r, expected);
+        }
+}
```
Comment (0xdaryl) [inline test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma_Strict.java]: Add EOL

File: test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma_Strict.java [hunk 7]
```diff
@@ -0,0 +1,549 @@
+/*******************************************************************************
+ * Copyright (c) 2001, 2019 IBM Corp. and others
```
Comment (0xdaryl) [inline test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma_Strict.java]: Same comment as above for the early copyright date.

File: test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+package org.openj9.test.java.lang;
+
+import org.testng.annotations.Test;
+import org.testng.AssertJUnit;
+
+@Test(groups={ "level.sanity" }, invocationCount=2)
+public class Test_Fma 
+    {
+    @Test
+	public void test_fma_double()
```
Comment (0xdaryl) [inline test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java]: Similar to the naming of the file, I think `test_Math_fma_double` to make the package explicit is a bit clearer.  Same comment applies to the Float and StrictMath test cases as well.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp (hunk 2 above)
Comment (wbh123456) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: I make them two separate functions because `inlineDoublePrecisionFPTrg1Src2` and `inlineSinglePrecisionFPTrg1Src2` functions. Do you think if it is a good idea to use a `bool` variable to indicate target register size if I combine the functions?

Comment (0xdaryl) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Why don't you ask the node for its data type and allocate the appropriate kind of targetRegister accordingly?  The targetRegister is what is set on the node, so it is dictated by the data type.

File: test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java (hunk 8 above)
Comment (wbh123456) [reply test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java]: Fixed in 1907302

File: test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma_Strict.java (hunk 7 above)
Comment (wbh123456) [reply test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma_Strict.java]: Fixed in 1907302

File: test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma_Strict.java (hunk 6 above)
Comment (wbh123456) [reply test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma_Strict.java]: Fixed in 1907302

File: test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java (hunk 5 above)
Comment (wbh123456) [reply test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java]: Fixed in 1907302

File: test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java (hunk 4 above)
Comment (wbh123456) [reply test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java]: Fixed in 1907302

Comment (wbh123456) [reply test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java]: Fixed in 1907302

File: test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java (hunk 3 above)
Comment (wbh123456) [reply test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Fma.java]: Fixed in 1907302

**Comment by zl-wang** [comment]
notice there is no single-precision vs. double-precision floating-point register distinction, especially on POWER.  you are  _not_  going to create different types of FPR for SP and DP. The fma opcode is different for SP and DP though. occasionally, you need frsp instruction to round DP into SP, but that depends on the particular situation.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp (hunk 2 above)
Comment (wbh123456) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Fixed in f980ee7

**Comment by wbh123456** [comment]
@0xdaryl I have fix the formatting issues and combine `inlineDoublePrecisionFPTrg1Src2` and `inlineSinglePrecisionFPTrg1Src2` into `inlineFPTrg1Src2`. I have also checked in the compilation log that correct instruction opcodes are used by associated Double/Singe fma method. This PR is ready for a round of review.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp [hunk 9]
```diff
@@ -9806,6 +9806,36 @@ static TR::Register *inlineSinglePrecisionFPTrg1Src2(TR::Node *node, TR::InstOpC
    return targetRegister;
    }
 
+static TR::Register *inlineFPTrg1Src3(TR::Node *node, TR::InstOpCode::Mnemonic op, TR::CodeGenerator *cg)
+   {
+   TR_ASSERT_FATAL(node->getNumChildren() == 3, "Wrong number of children in inlineFPTrg1Src3");
```
Comment (andrewcraik) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Can this assert be more helpful - can it state the number of children expected and the number of children found as well as the node number?

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp [hunk 10]
```diff
@@ -9806,6 +9806,36 @@ static TR::Register *inlineSinglePrecisionFPTrg1Src2(TR::Node *node, TR::InstOpC
...
+
+   TR::DataType type = node->getDataType();
+   TR_ASSERT_FATAL(type == TR::Float || type == TR::Double, "Wrong node DataType in inlineFPTrg1Src3");
```
Comment (andrewcraik) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Again can this be more helpful in telling you what went wrong?

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp (hunk 9 above)
Comment (wbh123456) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Fixed in 2f9a77

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp (hunk 10 above)
Comment (wbh123456) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Fixed in 2f9a77

**Comment by wbh123456** [comment]
@0xdaryl This PR is good for another review.

**Comment by wbh123456** [comment]
@0xdaryl Ping

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp [hunk 11]
```diff
@@ -12806,6 +12836,18 @@ J9::Power::CodeGenerator::inlineDirectCall(TR::Node *node, TR::Register *&result
             }
          break;
 
+      case TR::java_lang_Math_fma_D:
+      case TR::java_lang_StrictMath_fma_D:
+         resultReg = inlineFPTrg1Src3(node, TR::InstOpCode::fmadd, cg);
+         return true;
+         break;
```
Comment (0xdaryl) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Can you remove the `break` from here and the next case?  It is not needed, and it is inconsistent with the style of the other similar cases in this switch.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp [hunk 12]
```diff
@@ -9806,6 +9806,36 @@ static TR::Register *inlineSinglePrecisionFPTrg1Src2(TR::Node *node, TR::InstOpC
    return targetRegister;
    }
 
+static TR::Register *inlineFPTrg1Src3(TR::Node *node, TR::InstOpCode::Mnemonic op, TR::CodeGenerator *cg)
+   {
+   TR_ASSERT_FATAL(node->getNumChildren() == 3, "There should be exactly 3 children in the node of inlineFPTrg1Src3");
```
Comment (0xdaryl) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Perhaps you could output the node address and the number of children actually found as part of the message to make this a bit more helpful.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp [hunk 13]
```diff
@@ -9806,6 +9806,36 @@ static TR::Register *inlineSinglePrecisionFPTrg1Src2(TR::Node *node, TR::InstOpC
...
+
+   TR::DataType type = node->getDataType();
+   TR_ASSERT_FATAL(type == TR::Float || type == TR::Double, "Node of inlineFPTrg1Src3 should be either TR::Float or TR::Double");
```
Comment (0xdaryl) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Perhaps you could output the node address as part of the message.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp [hunk 14]
```diff
@@ -9806,6 +9806,36 @@ static TR::Register *inlineSinglePrecisionFPTrg1Src2(TR::Node *node, TR::InstOpC
...
+   TR::Node *thirdChild = node->getThirdChild();
+   TR::Register *src1Register = cg->evaluate(firstChild);
+   TR::Register *src2Register = cg->evaluate(secondChild);
+   TR::Register *src3Register = cg->evaluate(thirdChild);
+   TR::Register *targetRegister;
+
+   if(type == TR::Float)
+      targetRegister = cg->allocateSinglePrecisionRegister();
+   else
+   //DataType == TR::Double
```
Comment (0xdaryl) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: I don't think this comment is necessary.  Please remove.

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk11

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp (hunk 11 above)
Comment (wbh123456) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Fixed in 05e6a71

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp (hunk 12 above)
Comment (wbh123456) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Fixed in 05e6a71

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp (hunk 13 above)
Comment (wbh123456) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Fixed in 05e6a71

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp (hunk 14 above)
Comment (wbh123456) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Fixed in 05e6a71

**Comment by wbh123456** [comment]
Passed all gold tests.

**Comment by wbh123456** [comment]
@0xdaryl This one is ready to merge.

File: test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Math_Fma.java [hunk 15]
```diff
@@ -0,0 +1,549 @@
...
+        /**
+         * Testing special case: 
+         * If any argument is NaN, the result is NaN.
+         */
+        double a, b, c, r;
+        a = Double.NaN;
+        b = Math.random(); 
+        c = Math.random();
+        r = Math.fma(a, b, c);
+        AssertJUnit.assertTrue(Double.isNaN(r));
```
Comment (fjeremic) [inline test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Math_Fma.java]: I'm a little concerned with the use of `random()` above. If this assert ever fires, do you think it would be useful to know the values of `a`, `b`, and `c` so that we can easily debug and reproduce what happened? We should consider using the `String` argument to `assertTrue` [1] to pass a failure message displaying what the values of the inputs were.

[1] http://junit.sourceforge.net/javadoc/org/junit/Assert.html#assertTrue(java.lang.String,%20boolean)

Comment (wbh123456) [reply test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Math_Fma.java]: It completely makes sense to display what the random numbers are when the assertion fails. My purpose of randomizing those numbers was to find corner cases that we are unaware of. But in this case, I think it is sufficient enough to just use constant numbers instead.

Comment (fjeremic) [reply test/functional/Java9andUp/src/org/openj9/test/java/lang/Test_Math_Fma.java]: If using random numbers I'd say we need to display the numbers used in the assert failure string. If using constant numbers then we don't need to.

**Comment by wbh123456** [comment]
Changed random `float`/`double` in fma test cases to constant `float`/`double`. Checked that the fma test cases are still performing correctly.

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk11

**Comment by 0xdaryl** [comment]
@andrewcraik, would you mind reviewing to see if your concerns were addressed and approving if so?

**Comment by andrewcraik** [comment]
The code changes look fine now. Are there any tests where rounding is applied to ensure that the rounding is accurate? I saw there were some random number tests, but they don't seem to be specially addressing the rounding behaviours?

**Comment by wbh123456** [comment]
> The code changes look fine now. Are there any tests where rounding is applied to ensure that the rounding is accurate? I saw there were some random number tests, but they don't seem to be specially addressing the rounding behaviours?

Added 3 new test case which invokes rounding problems. Also tested the fma method against new test cases under linux ppc64 jdk11.

**Comment by 0xdaryl** [comment]
@andrewcraik , if all your issues have been resolved please formally approve and we can launch the final CI test on this PR.

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk11

**Comment by 0xdaryl** [comment]
Jenkins test sanity aix jdk11

**Comment by 0xdaryl** [comment]
The Linux x86-64 build actually passed but failed for an infrastructure reason on termination.  The AIX failures are infrastructural and the broader OpenJ9 builds are currently plagued with AIX build machine issues.  Given that it passes on Power Linux and that it passed a previous AIX CI build I'm going to merge this.
